### ISSUE TITLE
[Compiler Optimization] Induction variable recurrence flexibility, fallible loop dominance, and bounds-check elimination in align_mir::byte_ranges

### ISSUE THREAD

**Issue by sanohiro** [sanohiro/align#1049]
# [Compiler Optimization] Induction variable recurrence flexibility, fallible loop dominance, and bounds-check elimination in align_mir::byte_ranges

## Summary

During optimization of the decode loop and `greedy` argmax selection in `align-llm` (`src/runtime_generation.align`), an in-depth audit of the MIR optimization pipeline (`align_mir::byte_ranges.rs`) and generated LLVM IR revealed four compiler bottlenecks that force runtime bounds checks and inhibit LLVM load coalescing / vectorization:

1. **Hardcoded zero-based induction requirement in `byte_ranges.rs:278`**: `align_mir::byte_ranges` aborts bounds check elimination if a loop induction variable starts at a non-zero literal (e.g. `mut token := 1`).
2. **Fallible in-loop checks breaking dominance proofs in `byte_ranges.rs:427`**: `arm_dominates` fails if there is any early return or error branch inside the loop body (e.g. NaN check `if (bits & ...) == ... { return Err(Error.Invalid) }`), leaving all subsequent array/slice loads guarded by redundant runtime bounds checks (`align_rt_range_fail`).
3. **Untracked separate induction variables**: When a loop uses both a counter (`mut i := 0; i = i + 1`) and a byte offset (`mut offset := 0; offset = offset + 4`), the compiler treats `offset` as an unknown variable rather than a scaled recurrence `i * 4`. This prevents range-check elimination on `offset` and prevents LLVM from recognizing that `slice.u32_le(offset)` and `slice.f32_le(offset)` access the identical address, emitting two distinct loads instead of a single 32-bit load + bitcast.
4. **Spurious lossy conversion warnings on masked byte casts**: `(val & 255) as u8` generates `warning: lossy conversion: u32 as u8 truncates the high bits — this is defined behavior, not an error`, despite the high 24 bits being provably 0 via static bitmask analysis.

---

## 1. Hardcoded Zero-Based Induction in `byte_ranges.rs:278`

In `crates/align_mir/src/byte_ranges.rs:278`:
```rust
// Only 0-based loops for now: i = 0 .. n
if !literal(op, 0) {
    return None;
}
```

### Real-world impact in `align-llm`:
When scanning vocabulary logits in `greedy`, token 0 is often initialized before the loop as `best = 0; best_value = logits.f32_le(0)`. The natural loop then iterates over remaining tokens:
```align
mut token := 1
loop {
  if token >= n_vocab { break }
  val := logits.f32_le(token * 4)
  ...
  token = token + 1
}
```
Because the initial value is `1` rather than `0`, `literal(op, 0)` immediately returns `None`. Consequently, **none** of the slice bounds checks inside the loop are eliminated, emitting a runtime branch to `align_rt_range_fail` on every single iteration across all 151,936 tokens.

### Proposed Improvement:
Generalize induction recognition to support constant base offsets:
- Allow initial value `base = literal(op, k)`.
- Prove range `[k, k + n)` does not exceed `slice.len()` (or `slice.len() / element_size`).

---

## 2. In-Loop Fallible Checks Break Dominance Proofs

In `crates/align_mir/src/byte_ranges.rs:427`:
```rust
if !arm_dominates(header.id, take_true, block) {
    continue;
}
```

### Real-world impact in `align-llm`:
In robust runtime code, logits must be checked for NaN/Inf before consumption:
```align
loop {
  if i >= n_vocab { break }
  offset := i * 4
  bits := logits.u32_le(offset)
  if (bits & 2139095040) == 2139095040 { return Err(Error.Invalid) }
  value := logits.f32_le(offset)
  ...
  i = i + 1
}
```
Because the `if (bits & ...) == ... { return Err(...) }` statement branches to an error exit block:
1. `take_true` branches out of the normal loop path.
2. `arm_dominates(header.id, take_true, block)` evaluates to `false` for the subsequent block containing `logits.f32_le(offset)`.
3. The compiler gives up on eliminating the bounds check for `logits.f32_le(offset)`, emitting a redundant bounds check and disabling LLVM vectorization.

### Proposed Improvement:
An early exit that diverges or returns an `Err` from the function does not compromise the validity of the dominated path in the loop continuation. If the loop condition already establishes that `i * 4 < len`, any point that has not exited early is still within bounds.

---

## 3. Untracked Separate Induction Variables

Idiomatic Align code frequently maintains a byte offset alongside the iteration index:
```align
mut i := 0
mut offset := 0
loop {
  if i >= n { break }
  u := logits.u32_le(offset)
  f := logits.f32_le(offset)
  ...
  offset = offset + 4
  i = i + 1
}
```
When compiled this way:
1. `offset` is not tracked by the bounds eliminator, so both `u32_le` and `f32_le` retain separate runtime range checks.
2. Because the range check branches could theoretically panic or have side effects, LLVM cannot prove that `logits.u32_le(offset)` and `logits.f32_le(offset)` can be coalesced.
3. Replacing `offset` with `i * 4` allowed LLVM to coalesce both reads into:
   ```llvm
   %val32 = load i32, ptr %ptr, align 4
   %valf32 = bitcast i32 %val32 to float
   ```
   yielding an 84.8% microbenchmark speedup (1,287 µs -> 196 µs on 151.9k vocab).

### Proposed Improvement:
Add linear induction variable recognition in MIR: recognize `offset = offset + C` paired with `i = i + 1` as a linear affine relation `offset = base + i * C`, enabling both bounds-check elimination and LLVM load coalescing.

---

## 4. Spurious Lossy Conversion Warnings on Masked Byte Casts

When writing byte decomposition in Align:
```align
dest[base] = (val & 255) as u8
dest[base + 1] = ((val >> 8) & 255) as u8
```
Align emits:
```
warning: lossy conversion: `u32 as u8` truncates the high bits — this is defined behavior, not an error
```

### Problem:
The bitwise AND with `255` (`0xFF`) guarantees that all bits outside `[0..7]` are 0. The cast to `u8` cannot lose information.

### Proposed Improvement:
In `align_sema`, perform known-bits / mask analysis on integer expressions. When casting an integer type to a narrower integer type (e.g. `u32 as u8`), suppress the lossy conversion warning if the source expression's statically known upper bits are zero (e.g. masked by `0xFF`).

**Comment by sanohiro** [comment]
The selected compiler repairs are implemented and merged in #1056: byte-range proofs now cover nonnegative starting indices and paired/scaled offset recurrences, and finalized known-bit facts suppress integer-cast warnings only when the conversion is proved lossless.

The eight loop shapes and rejection controls preserve exact admitting-edge dominance and all unproved reached traps. The audit also found that the zero-based scaled loop with a nonfinite early return already optimized away its byte checks; early return alone was not the missing proof. This does not promise ordered argmax SIMD or a change to floating-point semantics.
